[JIT] Implement may_contain_alias in FunctionSchema

ttanpcs opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81562
* #81561
* #81575
* #81518
* #81444
* #81439
* __->__ #81352

- Created may_contain_alias method in FunctionSchema to publicize more detailed aliasing information about inputs and outputs of a schema. This method returns whether the first argument may contain an alias to the second argument (ie if the first argument is a list[Tensor], it can contain an alias to the second argument of the second argument is Tensor(*)) and vice versa if bidirectional = true.
- Created helper methods are explained more thoroughly in detail in function_schema.h
-Tested may_contain_alias methods for basic functionality, bidirectional functionality, wildcard functionality and dual container functionality in test_schema_info.cpp.

Commit:
[JIT] Implement may_contain_alias in FunctionSchema

[ghstack-poisoned]

Commit:
Update on "[JIT] Implement may_contain_alias in FunctionSchema"


- Created may_contain_alias method in FunctionSchema to publicize more detailed aliasing information about inputs and outputs of a schema. This method returns whether the first argument may contain an alias to the second argument (ie if the first argument is a list[Tensor], it can contain an alias to the second argument of the second argument is Tensor(*)) and vice versa if bidirectional = true.
- Created helper methods are explained more thoroughly in detail in function_schema.h
-Tested may_contain_alias methods for basic functionality, bidirectional functionality, wildcard functionality and dual container functionality in test_schema_info.cpp.

[ghstack-poisoned]

Commit:
Update on "[JIT] Implement may_contain_alias in FunctionSchema"


- Created may_contain_alias method in FunctionSchema to publicize more detailed aliasing information about inputs and outputs of a schema. This method returns whether the first argument may contain an alias to the second argument (ie if the first argument is a list[Tensor], it can contain an alias to the second argument of the second argument is Tensor(*)) and vice versa if bidirectional = true.
- Created helper methods are explained more thoroughly in detail in function_schema.h
-Tested may_contain_alias methods for basic functionality, bidirectional functionality, wildcard functionality and dual container functionality in test_schema_info.cpp.

[ghstack-poisoned]

davidberard98 reviewed (commented):
nice, looks good! Just had one comment about naming & moving a header import...

davidberard98 commented on aten/src/ATen/core/function_schema.cpp:
```diff
@@ -16,6 +16,48 @@ std::vector<Argument> FunctionSchema::getCorrectList(SchemaArgType type) const {
+    containedTypes.insert(current);
+  }
+
+  return std::vector<TypePtr> (containedTypes.begin(), containedTypes.end());
```
very minor nit, can you remove the space between `...<TypePtr>` and `(containedTypes...` ? unless there's a reason for it

Commit:
Update on "[JIT] Implement may_contain_alias in FunctionSchema"


- Created may_contain_alias method in FunctionSchema to publicize more detailed aliasing information about inputs and outputs of a schema. This method returns whether the first argument may contain an alias to the second argument (ie if the first argument is a list[Tensor], it can contain an alias to the second argument of the second argument is Tensor(*)) and vice versa if bidirectional = true.
- Created helper methods are explained more thoroughly in detail in function_schema.h
-Tested may_contain_alias methods for basic functionality, bidirectional functionality, wildcard functionality and dual container functionality in test_schema_info.cpp.

[ghstack-poisoned]

Commit:
Update on "[JIT] Implement may_contain_alias in FunctionSchema"


- Created may_contain_alias method in FunctionSchema to publicize more detailed aliasing information about inputs and outputs of a schema. This method returns whether the first argument may contain an alias to the second argument (ie if the first argument is a list[Tensor], it can contain an alias to the second argument of the second argument is Tensor(*)) and vice versa if bidirectional = true.
- Created helper methods are explained more thoroughly in detail in function_schema.h
-Tested may_contain_alias methods for basic functionality, bidirectional functionality, wildcard functionality and dual container functionality in test_schema_info.cpp.

[ghstack-poisoned]

Gamrix commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -402,6 +409,12 @@ struct TORCH_API FunctionSchema {
 
+  // Returns whether lhs and rhs may alias directly or whether lhs/rhs are a container
+  // that may contain elements that alias the other argument.
+  // bidirectional = false only returns whether lhs may contain an alias of rhs
+  // while bidirectional = true returns both directions.
+  bool may_contain_alias(const SchemaArgument& lhs, const SchemaArgument& rhs, bool bidirectional = true) const;
+
```
What is the use case for `bidirectional = false` ?

goldenxuett replied:
It's there if you want more specific information on aliasing. For instance if you have a Tensor[] and a Tensor then may_contain_alias(Tensor[], Tensor, false) might return true, while may_contain_alias(Tensor, Tensor[], false) will never return true.

Gamrix commented on aten/src/ATen/core/function_schema.cpp:
```diff
@@ -16,6 +17,48 @@ std::vector<Argument> FunctionSchema::getCorrectList(SchemaArgType type) const {
+  // process all further level contained types.
+  while (!typeStack.empty()) {
+    TypePtr current = typeStack.top();
+    typeStack.pop();
+    if (!containedTypes.count(current)) {
+      for (const TypePtr& containedType : current->containedTypes()) {
+        typeStack.push(containedType);
+      }
+    }
+    containedTypes.insert(current);
+  }
```
Nit: we could change this to first check if containedTypes has the type, and only if it does not, insert and iterate through its child types. This would also allow us to use a Vector instead of a unordered_set and avoid the extra conversion.

However, there is also always the saying: Done is better than perfect.

davidberard98 reviewed (approved):
looks good!! (wait for tests first just in case)

Gamrix reviewed (approved):
Generally looks good :D  Left a few comments here or there.

goldenxuett commented:
@pytorchbot merge
